ipi/engine/motion/eda.py

```diff
@@ -0,0 +1,337 @@
+        self._sigma = depend_value(
+            name="sigma", value=sigma if sigma is not None else np.inf
+        )
+        self._Efield = depend_array(
+            name="Efield",
+            value=np.zeros(3, float),
+            func=lambda: np.zeros(3, float),
+        )
+        self._Eenvelope = depend_value(
+            name="Eenvelope",
+            value=1.0,
+            func=lambda: np.zeros(3, float),
+        )
+
+    def bind(self, eda, enstype):
+        self.enstype = enstype
+        self._mts_time = eda._mts_time
+
+        dep = [self._mts_time, self._peak, self._sigma]
+        self._Eenvelope = depend_value(
+            name="Eenvelope", value=1.0, func=self._get_Eenvelope, dependencies=dep
+        )
+
+        if enstype in EDA.integrators:
+            # dynamics is not driven --> add dependencies to the electric field
+            dep = [self._mts_time, self._amp, self._freq, self._phase, self._Eenvelope]
+            self._Efield = depend_array(
+                name="Efield",
+                value=np.zeros(3, float),
+                func=self._get_Efield,
+                dependencies=dep,
+            )
+        else:
+            # dynamics is not driven --> no dependencies for the electric field
+            self._Efield = depend_array(
+                name="Efield",
+                value=np.zeros(3, float),
+                func=lambda: np.zeros(3, float),
+            )
+        pass
+
+    def store(self, ef):
+        super(ElectricField, self).store(ef)
+        self.amp.store(ef.amp)
+        self.freq.store(ef.freq)
+        self.phase.store(ef.phase)
+        self.peak.store(ef.peak)
+        self.sigma.store(ef.sigma)
+        pass
+
+    def _get_Efield(self):
+        """Get the value of the external electric field (cartesian axes)"""
+        time = dstrip(self.mts_time)
+        if hasattr(time, "__len__"):
+            return np.outer(self._get_Ecos(time) * self.Eenvelope, self.amp)
+        else:
+            return self._get_Ecos(time) * self.Eenvelope * self.amp
+
+    def _Eenvelope_is_on(self):
+        return self.peak > 0.0 and self.sigma != np.inf
+
+    def _get_Eenvelope(self):
+        time = dstrip(self.mts_time)
+        """Get the gaussian envelope function of the external electric field"""
+        if self._Eenvelope_is_on():
+            x = time  # indipendent variable
+            u = self.peak  # mean value
+            s = self.sigma  # standard deviation
+            return np.exp(
+                -0.5 * ((x - u) / s) ** 2
+            )  # the returned maximum value is 1, when x = u
+        else:
+            return 1.0
+
+    def _get_Ecos(self, time):
+        """Get the sinusoidal part of the external electric field"""
+        return np.cos(self.freq * time + self.phase)
+
+
+dproperties(
+    ElectricField,
+    ["amp", "phase", "peak", "sigma", "freq", "Eenvelope", "Efield", "mts_time"],
+)
+
+
+class EDA:
```

Comment:
yes I think it'd be better to combine all these objects in one. why passing around three objects that have always to be used together?

Reply:
https://github.com/i-pi/i-pi/pull/309#discussion_r1517507406